x-pack/plugins/ml/public/application/jobs/new_job/job_from_lens/visualization_extractor.ts

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectReference } from '@kbn/core/public';
+import type { DataViewsContract } from '@kbn/data-views-plugin/public';
+
+import type {
+  Embeddable,
+  LensPublicStart,
+  LensSavedObjectAttributes,
+  FieldBasedIndexPatternColumn,
+  XYDataLayerConfig,
+  IndexPatternPersistedState,
+  IndexPatternLayer,
+  XYLayerConfig,
+} from '@kbn/lens-plugin/public';
+import { layerTypes } from '@kbn/lens-plugin/public';
+
+import { i18n } from '@kbn/i18n';
+
+import { JOB_TYPE } from '../../../../../common/constants/new_job';
+import { ErrorType } from '../../../../../common/util/errors';
+import {
+  getVisTypeFactory,
+  isCompatibleLayer,
+  hasIncompatibleProperties,
+  hasSourceField,
+  isTermsField,
+  isCompatibleSplitFieldType,
+  getMlFunction,
+  getJobsItemsFromEmbeddable,
+  createDetectors,
+} from './utils';
+
+type VisualizationType = Awaited<ReturnType<LensPublicStart['getXyVisTypes']>>[number];
+
+export interface LayerResult {
+  id: string;
+  layerType: typeof layerTypes[keyof typeof layerTypes];
+  label: string;
+  icon: VisualizationType['icon'];
+  isCompatible: boolean;
+  jobType: JOB_TYPE | null;
+  error?: ErrorType;
+}
+
+export class VisualizationExtractor {
+  constructor(private dataViewClient: DataViewsContract) {}
+
+  public async getResultLayersFromEmbeddable(
+    embeddable: Embeddable,
+    lens: LensPublicStart
```

Comment:
should it be passed to the constructor instead?

Reply:
Ideally yes, but it means we'd also need to pass `lens` to `QuickJobCreator` as that calls this function.
This would mean in ML we would need to add `lens` to our dependancy cache (and I'd rather not add anything more in there as it needs to be removed)

I settled on passing it to this function only.